src/main/java/com/mongodb/spark/sql/connector/connection/MongoClientCache.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.connection;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.jetbrains.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.client.MongoClient;
+
+import com.mongodb.spark.sql.connector.assertions.Assertions;
+
+/**
+ * A simple cache that handles the acquisition and release of {@code MongoClients}.
+ *
+ * <p>Handles the creation and closing of {@code MongoClients}. When a {@link MongoClient} is no
+ * longer used there is a grace period allowing the {@code MongoClient} to be reused. As Spark
+ * Executors can operate on multiple tasks this reduces the cost creating a MongoClient and
+ * connections.
+ */
+final class MongoClientCache {
```

Comment:
This code is much more simple than the previous connectors cache. It does use synchronization locking on acquisition, release and cleanup of clients. This shouldn't be impediment as they are cheap operations.

Reply:
> This shouldn't be an impediment as they are cheap operations.

The algorithmic complexity of the method `checkClientCache` is O(numberOfCachedItems), which while may be considered cheap in some scenarios, it does not seem like something one may expect to be cheap in the absolute majority of use cases. And other methods are blocked by this linear-time method.

I understand that the decision to make all methods synchronized makes the implementation simpler. But this does not explain why we use `ArrayList` and linear-time search instead of using `HashMap` and constant-time search when accessing the cache. This decision makes the time complexity of all methods O(numberOfCachedItems), which is worrisome given that they also block each other.

Given that mongo-spark is not meant to be used as a library, but as a plugin (or something like that), I expect we are free to use 3rd party dependencies. If this is the case, can we consider using https://github.com/ben-manes/caffeine to get a cache? It's what [Guava encourages to use](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/cache/CacheBuilder.html) instead of Guava's cache.

Reply:
I think taking an extra dependency is complex for this as it really is in the library arena. We don't have control over the dependencies used by Spark or the users programs passed into Spark.

o(n) shouldn't be an issue as the cache is expected to be small (less than 100, in reality less than 10). The size of the cache is limited to each worker for the lifetime of a task. Its job is to reduce the cost of creating MongoClients when mulitple tasks are sent to the same Spark worker.  

It would be quite unusual for multiple differing Mongo-Spark configs to be required on the same Spark worker, at the same time, competing for the same resources. The keep alives are short enough to reduce the size of queue to make this effective enough.  

However, it can be revisited in the future should feedback or usecases change.

Reply:
I understand this, and I am OK with the linear algorithmic complexity of the `MongoClientCache.release` method. It could have been made constant-time if it were to use the same releasing approach that is used in `DefaultConnectionPool`: decorate `MongoClient` such that its `close` method calls `MongoClientCacheItem.release()` and does not call `super.close()`. But this approach requires writing more code as it requires to implement a simple `MongoClient` decorator.

But I still do not understand why `ArrayList` is used instead of `HashMap` to hold cached objects. I tried to find a single benefit this provides, and I failed. This approach has, however, multiple drawbacks:

1. This approach requires writing more code because instead of simply doing `return map.computeIfAbsent(mongoClientFactory, () -> new MongoClientCacheItem(mongoClientFactory, keepAliveNanos)).acquire()`, the method `MongoClientCache.acquire` has to find an element in a list and modify the list if the element is not there, which results in 11 lines of code where 1 would be enough if `HashMap` were used.
2. This approach is surprising to a reader and makes it think hard about the reasons `ArrayList` was chosen over `HashMap`. For me this thinking was futile.
3. This approach results in higher algorithmic complexity. Since this is irrelevant for the expected use cases, having even a small benefit of preferring the current approach could potentially outweigh this specific downside, but I failed to find a single benefit of using `ArrayList` instead of `HashMap`.

Could you please help me to understand what are the benefits of using `ArrayList` here instead of `HashMap`?